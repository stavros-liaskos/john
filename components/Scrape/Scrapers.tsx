import ScrapeButton from './components/ScrapeButton';
import React from 'react';

const Scrapers = () => {
  return (
    <div className="relative flex lg:justify-center items-center mb-2 h-20 md:h-40 border-b-2 rr-border w-full">
      <ScrapeButton buttonText={'Connect Spotify'} musicService={'Spotify'} iconName={'Spotify'} />
      <ScrapeButton buttonText={'Connect LastFm'} musicService={'LastFm'} iconName={'LastFm'} />
    </div>
  );
};

export default Scrapers;
