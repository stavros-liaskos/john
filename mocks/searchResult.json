{
  "artistsPerResource": {
    "fromDb": [],
    "fromLastfm": [
      {
        "id": "null",
        "lastfmUri": "https://www.last.fm/music/Liedfett",
        "name": "Liedfett",
        "spotifyUri": ""
      },
      {
        "id": "null",
        "lastfmUri": "https://www.last.fm/music/Leaders+of+the+New+School",
        "name": "Leaders of the New School",
        "spotifyUri": ""
      },
      {
        "id": "null",
        "lastfmUri": "https://www.last.fm/music/lo+&+leduc",
        "name": "lo & leduc",
        "spotifyUri": ""
      },
      {
        "id": "null",
        "lastfmUri": "https://www.last.fm/music/Luedji+Luna",
        "name": "Luedji Luna",
        "spotifyUri": ""
      },
      {
        "id": "null",
        "lastfmUri": "https://www.last.fm/music/Lead",
        "name": "Lead",
        "spotifyUri": ""
      },
      {
        "id": "null",
        "lastfmUri": "https://www.last.fm/music/Animals+as+Leaders",
        "name": "Animals as Leaders",
        "spotifyUri": ""
      },
      {
        "id": "null",
        "lastfmUri": "https://www.last.fm/music/Elfen+Lied",
        "name": "Elfen Lied",
        "spotifyUri": ""
      },
      {
        "id": "null",
        "lastfmUri": "https://www.last.fm/music/Lead+Belly",
        "name": "Lead Belly",
        "spotifyUri": ""
      },
      {
        "id": "null",
        "lastfmUri": "https://www.last.fm/music/Bay+Ledges",
        "name": "Bay Ledges",
        "spotifyUri": ""
      },
      {
        "id": "null",
        "lastfmUri": "https://www.last.fm/music/Ledapple",
        "name": "Ledapple",
        "spotifyUri": ""
      },
      {
        "id": "null",
        "lastfmUri": "https://www.last.fm/music/Leon+Leiden",
        "name": "Leon Leiden",
        "spotifyUri": ""
      },
      {
        "id": "null",
        "lastfmUri": "https://www.last.fm/music/Ledger",
        "name": "Ledger",
        "spotifyUri": ""
      },
      {
        "id": "null",
        "lastfmUri": "https://www.last.fm/music/Kiana+Led%C3%A9",
        "name": "Kiana Led√©",
        "spotifyUri": ""
      },
      {
        "id": "null",
        "lastfmUri": "https://www.last.fm/music/Monsters+of+Liedermaching",
        "name": "Monsters of Liedermaching",
        "spotifyUri": ""
      },
      {
        "id": "null",
        "lastfmUri": "https://www.last.fm/music/Lendakaris+Muertos",
        "name": "Lendakaris Muertos",
        "spotifyUri": ""
      },
      {
        "id": "null",
        "lastfmUri": "https://www.last.fm/music/Kinder+Lieder",
        "name": "Kinder Lieder",
        "spotifyUri": ""
      },
      {
        "id": "null",
        "lastfmUri": "https://www.last.fm/music/Led+Er+Est",
        "name": "Led Er Est",
        "spotifyUri": ""
      },
      {
        "id": "null",
        "lastfmUri": "https://www.last.fm/music/Cult+Leader",
        "name": "Cult Leader",
        "spotifyUri": ""
      },
      {
        "id": "null",
        "lastfmUri": "https://www.last.fm/music/Jefre+Cantu-Ledesma",
        "name": "Jefre Cantu-Ledesma",
        "spotifyUri": ""
      },
      {
        "id": "null",
        "lastfmUri": "https://www.last.fm/music/Arlette+Leduc",
        "name": "Arlette Leduc",
        "spotifyUri": ""
      },
      {
        "id": "null",
        "lastfmUri": "https://www.last.fm/music/Leadley",
        "name": "Leadley",
        "spotifyUri": ""
      },
      {
        "id": "null",
        "lastfmUri": "https://www.last.fm/music/Tomas+Ledin",
        "name": "Tomas Ledin",
        "spotifyUri": ""
      },
      {
        "id": "null",
        "lastfmUri": "https://www.last.fm/music/Chris+LeDoux",
        "name": "Chris LeDoux",
        "spotifyUri": ""
      },
      {
        "id": "null",
        "lastfmUri": "https://www.last.fm/music/Leadbelly",
        "name": "Leadbelly",
        "spotifyUri": ""
      },
      {
        "id": "null",
        "lastfmUri": "https://www.last.fm/music/Ledisi",
        "name": "Ledisi",
        "spotifyUri": ""
      },
      {
        "id": "null",
        "lastfmUri": "https://www.last.fm/music/Lewd+Acts",
        "name": "Lewd Acts",
        "spotifyUri": ""
      },
      {
        "id": "null",
        "lastfmUri": "https://www.last.fm/music/Letdown.",
        "name": "Letdown.",
        "spotifyUri": ""
      },
      {
        "id": "null",
        "lastfmUri": "https://www.last.fm/music/Led+Zeppelin",
        "name": "Led Zeppelin",
        "spotifyUri": ""
      },
      {
        "id": "null",
        "lastfmUri": "https://www.last.fm/music/Led+Zepagain",
        "name": "Led Zepagain",
        "spotifyUri": ""
      },
      {
        "id": "null",
        "lastfmUri": "https://www.last.fm/music/Future+Leaders+of+the+World",
        "name": "Future Leaders of the World",
        "spotifyUri": ""
      }
    ]
  }
}
