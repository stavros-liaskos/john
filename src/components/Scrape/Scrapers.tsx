import ScrapeButton from './components/ScrapeButton';
import React from 'react';

const Scrapers = () => {
  return (
    <div className="flex lg:justify-center gap-2 mb-2 w-full">
      <ScrapeButton buttonText={'Connect'} musicService={'Spotify'} iconName={'Spotify'} />
      <ScrapeButton buttonText={'Connect'} musicService={'LastFm'} iconName={'LastFm'} />
    </div>
  );
};

export default Scrapers;
