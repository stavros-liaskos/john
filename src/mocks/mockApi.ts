import nock from 'nock';
import followedArtists from './fixtures/responses/followed-artists.json';
import { components } from '../types/schema';

export const nockAuth = {
  success: () => nockAPI().get(`/api/auth/me`).reply(200),
};

export const nockFollowedArtists = {
  success: (artistQuantity: number = 2) => {
    followedArtists.rows.splice(1, 2 - artistQuantity);
    nockAPI().get('/me/followed-artists').reply(200, followedArtists);
  },
};

export const nockRecommendedArtists = {
  success: () => nockAPI().get('/artists/recommended').query({ page: 0, size: 10 }).reply(200, followedArtists),
};

export const nockScrape = {
  success: () => nockAPI().get('/scrape-taste/spotify').reply(200, {}),
  fail: () => nockAPI().get('/scrape-taste/spotify').reply(200),
};

export const nockFollow = {
  success: () => nockAPI().post('/me/follow').reply(200),
};

export const nockUnfollow = {
  success: () => nockAPI().delete('/me/unfollow').reply(200),
};

export const nockSearch = {
  success: (res: components['schemas']['ArtistSearchResponse'], searchQuery: string) =>
    nockAPI().get('/artist/search').query({ pattern: searchQuery }).reply(200, res),
};

function nockAPI() {
  return nock(process.env.BE_BASE_URL!);
}
