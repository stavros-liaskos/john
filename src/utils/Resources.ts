import { components } from '../types/schema';

interface IResources {
  cacheMap: Record<string, unknown>;
  fetch: (url: string) => void;
}

class Resources implements IResources {
  private cacheMap;

  constructor() {
    this.cacheMap = new Map();
  }

  fetch(url) {
    if (this.cacheMap.has(url)) {
      return this.cacheMap.get(url);
    }

    const promise = fetch(url, {
      method: 'GET',
      credentials: 'include',
    })
      .then(response => {
        return response.json();
      })
      .then((data: components['schemas']['FollowedArtistsResponse']) => {
        this.cacheMap.set(url, data);
        return data;
      });

    throw promise;
  }
}

const resources = new Resources();

export default resources;
