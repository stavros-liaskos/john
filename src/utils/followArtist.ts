import { components } from '../types/schema';
import Endpoints from '../types/endpoints';

async function followArtist(artist: components['schemas']['SearchResultArtistDto'], cb: () => void) {
  const headers = new Headers({ 'Content-Type': 'application/json' });
  await fetch(Endpoints.Follow, {
    method: 'POST',
    credentials: 'include',
    headers,
    body: JSON.stringify(artist),
  })
    .then(() => {
      console.log(`${artist.name} followed successfully`);
    })
    .catch(error => {
      console.error('Error:', JSON.stringify(error));
    })
    .finally(() => {
      cb();
    });
}
export default followArtist;
