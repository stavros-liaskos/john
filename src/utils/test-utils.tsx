import React from 'react';
import { render, act } from '@testing-library/react';
import ThemeProvider from '../contexts/Theme/ThemeProvider';
import { UserProvider } from '@auth0/nextjs-auth0/client';
import ArtistsListProvider from '../contexts/ArtistsList/ArtistsListProvider';

const AllTheProviders = ({ children }: { children: React.ReactNode }) => {
  return (
    <ThemeProvider>
      <UserProvider>
        <ArtistsListProvider>{children}</ArtistsListProvider>
      </UserProvider>
    </ThemeProvider>
  );
};

const customRender = (ui: React.ReactElement, options?: { [key: string]: unknown }) =>
  render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render };

export const renderWithAct = async (ui: React.ReactElement, options?: { [key: string]: unknown }) => {
  const component = render(ui, { wrapper: AllTheProviders, ...options });

  const promise = Promise.resolve();
  await act(() => promise);

  return component;
};

export const beforeEachTest = () => {
  jest.resetModules();
  jest.clearAllMocks();
  jest.restoreAllMocks();
  localStorage.clear();
};
