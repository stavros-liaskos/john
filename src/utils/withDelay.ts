export default function withDelay(cb: () => void, delay = 1000) {
  return setTimeout(() => {
    cb();
  }, delay);
}
