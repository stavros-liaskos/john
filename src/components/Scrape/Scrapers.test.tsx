import Scrapers from './Scrapers';
import { setupServer } from 'msw/node';
import { beforeEachTest, renderWithAct } from '../../utils/test-utils';
import { mswAuth, mswFollowedArtists, mswRaccoonUser } from '../../mocks/mockApi';

describe('Scrapers', () => {
  const server = setupServer();

  beforeAll(() => {
    server.listen();
    server.listen({
      onUnhandledRequest: 'error',
    });
  });
  afterEach(() => {
    jest.restoreAllMocks();
    server.resetHandlers();
  });
  afterAll(() => server.close());

  beforeEach(() => {
    beforeEachTest();
    server.use(mswFollowedArtists.success(2), mswAuth.success(), mswRaccoonUser.success());
  });

  it('loads spotify in "Scraped" state and lastfm in "Not scraped"', async () => {
    const { findAllByRole } = await renderWithAct(<Scrapers />);

    expect(await findAllByRole('img')).toHaveLength(2);
  });

  it('matches snapshot', async () => {
    const { container } = await renderWithAct(<Scrapers />);

    expect(container).toMatchSnapshot();
  });
});
