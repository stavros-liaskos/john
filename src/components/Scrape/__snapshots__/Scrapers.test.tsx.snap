// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scrapers matches snapshot 1`] = `
<div>
  <div
    class="flex lg:justify-center gap-2 mb-2 w-full"
  >
    <div
      class="flex justify-center items-center mb-2 w-full"
    >
      <button
        class="btn flex flex justify-between py-2 px-3 w-full md:w-48 !rr-text-confirm"
        disabled=""
        type="button"
      >
        Connected
        <svg
          height="30"
          role="img"
          viewBox="0 0 128 128"
          width="30"
          xmlns="http://www.w3.org/2000/svg"
        >
          <g
            clip-rule="evenodd"
          >
            <path
              d="M0 0h128v128H0z"
              fill="none"
            />
            <path
              d="M64 0C28.8 0 0 28.8 0 64s28.8 64 64 64 64-28.8 64-64S99.52 0 64 0m29.44 92.48c-1.28 1.92-3.52 2.56-5.44 1.28-15.04-9.28-33.92-11.2-56.32-6.08-2.24.64-4.16-.96-4.8-2.88-.64-2.24.96-4.16 2.88-4.8 24.32-5.44 45.44-3.2 62.08 7.04 2.24.96 2.56 3.52 1.6 5.44m7.68-17.6c-1.6 2.24-4.48 3.2-6.72 1.6-17.28-10.56-43.52-13.76-63.68-7.36-2.56.64-5.44-.64-6.08-3.2-.64-2.56.64-5.44 3.2-6.08 23.36-7.04 52.16-3.52 72 8.64 1.92.96 2.88 4.16 1.28 6.4m.64-17.92C81.28 44.8 47.04 43.52 27.52 49.6c-3.2.96-6.4-.96-7.36-3.84-.96-3.2.96-6.4 3.84-7.36 22.72-6.72 60.16-5.44 83.84 8.64 2.88 1.6 3.84 5.44 2.24 8.32-1.6 2.24-5.44 3.2-8.32 1.6"
              fill="#10BC4C"
              fill-rule="evenodd"
            />
          </g>
        </svg>
      </button>
    </div>
    <div
      class="flex justify-center items-center mb-2 w-full"
    >
      <button
        class="btn flex flex justify-between py-2 px-3 w-full md:w-48"
        type="button"
      >
        Connect
        <svg
          height="30"
          role="img"
          viewBox="0 0 112.195 112.195"
          width="30"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M112.195 56.097c0 30.983-25.114 56.099-56.097 56.099C25.115 112.195 0 87.08 0 56.097 0 25.117 25.115 0 56.099 0c30.982 0 56.096 25.117 56.096 56.097z"
            fill="#d31f27"
          />
          <path
            d="M80.487 52.141c-6.203-1.849-8.412-2.874-8.412-4.922 0-3.416 4.937-4.702 5.497-4.79 3.139-.465 7.072 1.173 8.34 4.99l8.44-2.553c-2.65-8.385-10.487-12.099-18.099-10.975-7.788 1.146-12.782 6.786-12.782 13.327 0 8.8 8.244 11.332 14.43 13.181 6.509 1.944 8.401 2.637 8.401 5.023 0 2.016-1.196 3.895-4.374 4.735-6.32 1.664-14.586-.672-16.887-4.523-1.351-2.261-2.733-5.203-4.001-8.53-3.766-9.866-8.926-23.371-23.764-23.371-7.731 0-21.851 3.743-21.851 24.868 0 8.744 7.481 19.613 21.578 19.613 11.829 0 13.888-4.32 14.358-5.085L47.5 65.574c-.098.138-3.692 4.936-10.497 4.936-11.107 0-12.735-11.776-12.735-11.909 0-10.967 4.687-16.238 13.008-16.238 8.094 0 11.382 7.04 15.48 17.759 1.397 3.683 2.922 6.966 4.65 9.865 3.657 6.116 11.55 8.091 19.473 8.091 2.418 0 5.2.137 7.351-.432 7.093-1.878 10.905-6.178 10.905-12.224.002-9.204-8.283-11.378-14.648-13.281z"
            fill="#fff"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;
